test/broadway_test.exs

```diff
@@ -441,6 +457,66 @@ defmodule BroadwayTest do
     end
   end
 
+  describe "handle producer crash" do
+    setup do
+      test_pid = self()
+
+      handle_message = fn message, _ ->
+        send(test_pid, {:message_handled, message})
+        message
+      end
+
+      context = %{
+        handle_message: handle_message,
+        handle_batch: fn _, batch, _, _ -> batch end
+      }
+
+      broadway_name = new_unique_name()
+
+      {:ok, _pid} =
+        Broadway.start_link(ForwarderWithCustomHandlers,
+          name: broadway_name,
+          resubscribe_interval: 0,
+          context: context,
+          producers: [
+            default: [
+              module: ManualProducer,
+              arg: %{test_pid: self()},
+              transformer: {Transformer, :transform, []}
+            ]
+          ],
+          processors: [stages: 1, min_demand: 1, max_demand: 2],
+          publishers: [default: [batch_size: 2]]
+        )
+
+      :ok
+    end
+
+    @tag :capture_log
+    test "producer will be restarted" do
+      assert_receive {:producer_initialized, producer}
+      refute_receive {:producer_initialized, _}
+
+      async_push_messages(producer, [:kill_producer])
+
+      assert_receive {:producer_initialized, ^producer}
+    end
+
+    @tag :capture_log
+    test "processors resubscribe to the restarted producers and keep processing messages" do
+      assert_receive {:producer_initialized, producer}
+
+      push_messages(producer, [1])
+      assert_receive {:message_handled, %{data: 1}}
+
+      async_push_messages(producer, [:kill_producer])
```

Comment:
Any reason why you are using the async version? The async version is a bit hackish. The regular version should be enough, no?

Reply:
`push_messages` is a call, so when I kill the receiver process during that call, it brings down the caller too, which makes the test fail. Can I test it differently? Not sure how.

Reply:
Ah, good point!

Reply:
Actually, I think maybe we should not use push_messages on this test. I proposed another approach here: https://github.com/plataformatec/broadway/pull/28/files#r255296274